src/player.cpp

```diff
@@ -5156,24 +5156,35 @@ void player::suffer()
         }
     } // Done with while-awake-only effects
 
-    if( has_trait( trait_ASTHMA ) && one_in(3600 - stim * 50) &&
+    if( has_trait( trait_ASTHMA ) && one_in( ( 3600 - stim * 50 ) * ( has_effect( effect_sleep ) ? 10 : 1 ) ) &&
         !has_effect( effect_adrenaline ) & !has_effect( effect_datura ) ) {
-        bool auto_use = has_charges("inhaler", 1);
-        if (underwater) {
+        bool auto_use = has_charges( "inhaler", 1 );
+        if ( underwater ) {
             oxygen = oxygen / 2;
             auto_use = false;
         }
 
         if( has_effect( effect_sleep ) ) {
-            add_msg_if_player(_("You have an asthma attack!"));
-            wake_up();
-            auto_use = false;
-        } else {
             add_msg_if_player( m_bad, _( "You have an asthma attack!" ) );
-        }
-
-        if (auto_use) {
-            use_charges("inhaler", 1);
+            inventory map_inv;
+            map_inv.form_from_map( g->u.pos(), 2 );
```

Comment:
Why limit it to 2 tiles? We have `PICKUP_RANGE` that equals to 6 tiles.

Reply:
This is supposed to be done by half-asleep character and the inhaler must be placed near the bed. Searching for it at longer range will require awakening. For the same reason I did not included oxygen tanks because they are more complicated to use.

Reply:
> This is supposed to be done by half-asleep character and the inhaler must be placed near the bed

Makes sense.